packages/replay-next/src/suspense/AnalysisCache.ts

```diff
@@ -81,59 +66,83 @@ export function createAnalysisCache<
       const params = paramsWithCacheLoadOptions.slice(0, -1) as TParams;
-      const analysisParams = createAnalysisParams(...params);
-      const locations = analysisParams.location
-        ? client.getCorrespondingLocations(analysisParams.location).map(location => ({
-            location,
-          }))
-        : undefined;
-      if (locations) {
+      const points = await findPoints(client, begin, end, ...params);
+      onPointsReceived?.(points);
+
+      const evaluationParams = createEvaluationParams(points, ...params);
+      let allEvaluationParams = [evaluationParams];
+      if (evaluationParams.selector.kind === "location") {
+        const locations = client.getCorrespondingLocations(evaluationParams.selector.location);
+        allEvaluationParams = locations.map(location => ({
+          ...evaluationParams,
+          location,
+        }));
         await Promise.all(
-          locations.map(location =>
-            breakpointPositionsCache.readAsync(client, location.location.sourceId)
-          )
+          locations.map(location => breakpointPositionsCache.readAsync(client, location.sourceId))
         );
       }
 
-      let allPoints: TPoint[] = [];
-      let error: any;
-      await client.streamAnalysis(
-        {
-          locations,
-          eventHandlerEntryPoints: analysisParams.eventTypes?.map(eventType => ({ eventType })),
-          exceptionPoints: analysisParams.exceptions,
-          mapper: analysisParams.mapper,
-          effectful: false,
-          range: { begin, end },
-        },
-        {
-          onPoints: points => {
-            allPoints = allPoints.concat(points.map(point => transformPoint(point, ...params)));
+      for (const evaluationParams of allEvaluationParams) {
+        client.runEvaluation(
+          {
+            selector: evaluationParams.selector,
+            expression: evaluationParams.expression,
+            frameIndex: evaluationParams.frameIndex,
+            fullPropertyPreview: evaluationParams.fullPropertyPreview,
+            limits: { begin, end },
           },
-          onResults: analysisEntries => {
-            for (const analysisEntry of analysisEntries) {
-              const result = analysisEntry.value as RemoteAnalysisResult;
+          async results => {
+            for (const result of results) {
+              setPointAndTimeForPauseId(result.pauseId, result.point);
               cachePauseData(client, result.pauseId, result.data);
 
-              resultsCache.cache(result, result.point, ...params);
-            }
-          },
-          onError: err => (error = err),
-        }
-      ).pointsFinished;
+              const values: Value[] = [];
+              if (result.exception) {
+                values.push(result.exception);
+              } else if (result.returned?.object) {
+                const length =
+                  (
+                    await objectPropertyCache.readAsync(
```

Comment:
Does cache code like this trigger any protocol commands or does this just query the cache for the value?

Reply:
It potentially does both. If the value is already cached, it will return synchronously. Else it will query `Pause.getObjectProperty`.

Reply:
Looking at the code here, these _may_ have come back as part of the `results` handler on line 99?  Would need to run this / add some logging and see.

Pretty sure this _is_ necessary, though. The whole point of this file is to take whatever evaluation code was provided (such as a print statement content), eval that, and show it.  This chunk appears to be taking those results.

In other words, if a print statement was:

`"myFunction", state.value, counter`

We convert that into an array like `["myFunction", state.value, counter]`, and what's coming back here _is_ the array we need to show the results in the console.

But, we want to show each value individually in the console message, not wrapped in that artificial array.  So, we have to extract the values here, which means doing the array lookup to get them index-by-index.

(This is yet another example of how working with the protocol's design is frequently a pain.)

Reply:
If this code is going to live here, it should only be used in cases where `fullPropertyPreview: true` is passed, since it's designed to address that case as much as possible. That said, either way the client should still be getting back a `canOverflow` preview vs a `full` preview. Is the frontend required to have a `full`-level preview in order to render the object?

>   So, we have to extract the values here, which means doing the array lookup to get them index-by-index.

My concern is that they should already be in the cache, just potentially with a non `full` preview level. If they're not there at all, then something about that object preview has probably gone wrong.

What I'm saying for "user interaction" is primarily that from a perf standpoint:
* It'd be _way_ better to show the `canOverflow` preview and then have the `full` preview loading in the background, rather than `await`-ing the preview here.
* It'd be even better to just show the `canOverflow` preview, and only query the `full` preview if the user clicks on the object in some way. The `fullPropertyPreview` will do its best to expand things from `canOverflow` to `full` as part of `runEvaluation`, so if the frontend can just show the `full` when it's available and `canOverflow` overwise (until the user interacts), that would be ideal

Reply:
I feel like we're maybe talking past each other a bit here?

In general, the console message UI for print statements is wanting to do something akin to `values.map(value => <ObjectInspector value={value} />`, and then the inspector takes care of asking for more details.

I'm not sure of the specifics of overflow vs non-overflow currently.

We have other cases for analysis as well, but that's the main one where this matters, I think.

Reply:
Sorry, I'm a few messages behind. Been in meetings.

> Hmm okay, I really don't think we should be doing this then. The frontend should do its absolute best to only query pause objects based on user interaction. Pause commands can be really slow, especially in this case where it's being used for the first time.

The frontend generally does this, only querying what it needs to display the data. Our caching supports 3 tiers of "preview" data– none, can-overflow, and full.

API requests made to `Pause.getObjectPreview` and `Pause.getObjectProperty` both cache data for a given Pause ID so we avoid re-requesting data unless (a) we need it to render something and (b) it hasn't already been cached as part of a prior request. (Indeed if we _forget_ this, the backend won't re-send the data anyway– so this cache is really important).

I'm not familiar with the new Analysis API (or the changes in this PR really) but given the call to `cachePauseData ` a few lines above:
https://github.com/replayio/devtools/blob/561d3ec018367fdfb16e12cde58875bbc38182ba/packages/replay-next/src/suspense/AnalysisCache.ts#LL102-L102C67

...I suspect the call you're commenting on (to `await objectPropertyCache.readAsync`) would be a **cache hit** in most cases.

Reply:
Yeah I think we probably are a bit. Re-reading this code again, I misunderstood one thing, which is that this is more focused on the `ProtocolValue` than it is on the `ObjectPreview`.

Given that, I think the only comment I have left is that if these values aren't already in the cache, I'd consider that very bad and an indication that something's gone wrong, so because this is using the async version of the function, I'm worried that it could mask that problem by instead issuing further pause commands.

Reply:
If that's the case (I don't have much context on this cache instance) we could replace–
```js
await objectPropertyCache.readAsync(...)
```
with–
```js
 await objectPropertyCache.getValue(...)
```
The latter function will return a cached value _or thrown_ if it's a cache miss.

Reply:
> Looking at the code here, these may have come back as part of the results handler on line 99?

Exactly.

> If this code is going to live here, it should only be used in cases where fullPropertyPreview: true is passed

That's exactly what's happening here (only in this case it's the `fullReturnedPropertyPreview` param for `Session.runEvaluation`). In my tests, all necessary previews were already cached, so this didn't send a single backend request.

Reply:
>> If this code is going to live here, it should only be used in cases where fullPropertyPreview: true is passed
>
> That's exactly what's happening here

We could ensure this remains the case going forward by replacing `readAsync` with `getValue`.

Is there a reason not to do this, @hbenl ?

Reply:
> Is there a reason not to do this

I think we don't have a guarantee that the backend will include the necessary object previews in the events triggered by `Session.runEvaluation` and if it decides to not send an object preview for whatever reason, I'd prefer the frontend to request that rather than failing to show the message or showing an incomplete message.